File: zemiao/274.py
class Solution(object):
    def hIndex(self, citations):
        """
        :type citations: List[int]
        :rtype: int
        """
        n = len(citations)
        
        # buckets[i] is the count of citations >= i
        buckets = [0]*(n+1)
        for i in citations:
          if 0 < i <= n:
            buckets[i] += 1
          elif i > n:
            buckets[n] += 1
        
        for i in reversed(range(1,n)):
          buckets[i] += buckets[i+1]
        
        res = 0
        for i in range(1,n+1):
          if i <= buckets[i]:
            res = i
        return res

File: tnzhou/274.cpp
class Solution {
public:
    int hIndex(vector<int>& citations) {
        sort(citations.begin(), citations.end());
        if (citations.size() == 0)
            return 0;
        int res;
        int left = 0;
        int right = citations.size();
        while (left <= right) {
            int m = (left + right) / 2;
            if (citations[m] >= citations.size() - m) {
                res = citations.size() - m;
                right = m - 1;
            } else {
                left = m + 1 ;
            }
        }
        return res;
    }
};
